warplib/src/warp_memory.cairo

```diff
@@ -14,24 +25,70 @@ impl DestructWarpMemory of Destruct::<WarpMemory> {
 }
 
 
-trait MemoryTrait {
+trait WarpMemoryTrait {
     fn initialize() -> WarpMemory;
-    fn insert(ref self: WarpMemory, position: felt252, value: felt252);
-    fn append(ref self: WarpMemory, value: felt252);
+
+    fn unsafe_write(ref self: WarpMemory, position: felt252, value: felt252);
+    fn write(ref self: WarpMemory, position: felt252, value: felt252);
+
+    fn unsafe_read(ref self: WarpMemory, position: felt252) -> felt252;
+    fn read(ref self: WarpMemory, position: felt252) -> felt252;
+
+    /// Given a certain size, it allocates the space for writing
+    fn unsafe_alloc(ref self: WarpMemory, size: felt252) -> felt252;
+    fn alloc(ref self: WarpMemory, size: felt252) -> felt252;
 }
 
 
-impl WarpMemoryImpl of MemoryTrait {
+impl WarpMemoryImpl of WarpMemoryTrait {
     fn initialize() -> WarpMemory {
         return WarpMemory {memory: Felt252DictTrait::new(), pointer: 0};
     }
 
-    fn insert(ref self: WarpMemory, position: felt252, value: felt252) {
+    fn unsafe_write(ref self: WarpMemory, position: felt252, value: felt252){
+        self.memory.insert(position, value);
+    }
+
+    fn write(ref self: WarpMemory, position: felt252, value: felt252){
+        let position_256 = u256_from_felt252(position);
+        let pointer_256 = u256_from_felt252(self.pointer);
+        if position_256 >= pointer_256 {
+           panic_with_felt252('Writing on unreserved position');
+        }
         self.memory.insert(position, value);
-        self.pointer += 1;
     }
 
-    fn append(ref self: WarpMemory, value: felt252) {
-        self.insert(self.pointer, value);
+    fn unsafe_read(ref self: WarpMemory, position: felt252) -> felt252 {
+        self.memory.get(position)
+    }
+
+    fn read(ref self: WarpMemory, position: felt252) -> felt252 {
+        let position_256 = u256_from_felt252(position);
+        let pointer_256 = u256_from_felt252(self.pointer);
+        if position_256 >= pointer_256 {
+            panic_with_felt252('Reading on unreserved position');
+        }
+        self.memory.get(position)
+    }
+
+    fn unsafe_alloc(ref self: WarpMemory, size: felt252) -> felt252 {
+        let reserved_pointer = self.pointer;
+        self.pointer += size;
+        
+        reserved_pointer
+    }
+
+    fn alloc(ref self: WarpMemory, size: felt252) -> felt252 {
+        let pointer_256 = u256_from_felt252(self.pointer);
+        let size_256 = u256_from_felt252(size);
+
+        let MAX_FELT_256: u256 = u256_from_felt252(-1);
+        if (pointer_256 + size_256 > MAX_FELT_256){
```

Comment:
You can directly use `assert` function here

https://github.com/starkware-libs/cairo/blob/74dbf871c1125a591591519b75f987df1fb2d08d/corelib/src/lib.cairo#L391

Reply:
there's other condition in the same script that can be replaced

Reply:
yes using assert is more readable. In this specific case where we are dealing with functions that have a high probability of being called  a lot, reducing an extra function call is noticeable performance wise